community/cypher/frontend-3.3/src/main/scala/org/neo4j/cypher/internal/frontend/v3_3/ExpressionStringifier.scala

```diff
@@ -0,0 +1,250 @@
+        val p = s.innerPredicate.map(e => " WHERE " + this.apply(e)).getOrElse("")
+        val e = s.extractExpression.map(e => " | " + this.apply(e)).getOrElse("")
+        val expr = this.apply(expression)
+        s"extract($v IN $expr$p$e)"
+      case PatternComprehension(variable, RelationshipsPattern(relChain), predicate, proj, _) =>
+        val v = variable.map(e => s"${this.apply(e)} = ").getOrElse("")
+        val p = predicate.map(e => " WHERE " + this.apply(e)).getOrElse("")
+        s"[$v${pattern(relChain)}$p | ${this.apply(proj)}]"
+      case e@HasLabels(arg, labels) =>
+        val l = labels.map(label => backtick(label.name)).mkString(":", ":", "")
+        s"${parens(e, arg)}$l"
+      case AllIterablePredicate(scope, e) =>
+        s"all${prettyScope(scope, e)}"
+      case NoneIterablePredicate(scope, e) =>
+        s"none${prettyScope(scope, e)}"
+      case MapProjection(variable, items, _) =>
+        val itemsText = items.map {
+          case LiteralEntry(k, e) => s"${backtick(k.name)}: ${this.apply(e)}"
+          case VariableSelector(v) => this.apply(v)
+          case PropertySelector(v) => s".${this.apply(v)}"
+          case AllPropertiesSelector() => ".*"
+        }.mkString(", ")
+        s"${this.apply(variable)}{$itemsText}"
+      case CaseExpression(expression, alternatives, default) =>
+        val e = expression.map(e => s" ${this.apply(e)}").getOrElse("")
+        val d = default.map(e => s" else ${this.apply(e)} ").getOrElse("")
+        val items = (alternatives map {
+          case (e1, e2) => s"when ${this.apply(e1)} then ${this.apply(e2)}"
+        }).mkString(" ", " ", "")
+        s"case$e$items${d}end"
+      case e@Ands(expressions) =>
+        expressions.map(x => parens(e, x)).mkString(" AND ")
+      case e@Ors(expressions) =>
+        expressions.map(x => parens(e, x)).mkString(" OR ")
+      case ShortestPathExpression(s@ShortestPaths(r:RelationshipChain, _)) =>
+        s"${s.name}(${pattern(r)})"
+      case _: ExtractScope | _: FilterScope | _: ReduceScope =>
+        // These are not really expressions, they are part of expressions
+        ""
+      case _ =>
+        println(s"$ast failed on default, trying extender")
+        extender(ast)
+    }
+  }
+
+  private def parens(caller: Expression, argument: Expression) = {
+    val thisPrecedence = precedenceLevel(caller)
+    val argumentPrecedence = precedenceLevel(argument)
+    if (argumentPrecedence >= thisPrecedence)
+      s"(${this.apply(argument)})"
+    else
+      this.apply(argument)
+  }
+
+  private def prettyScope(s: FilterScope, expression: Expression) = {
+    val v = this.apply(s.variable)
+    val e = this.apply(expression)
+    val p = s.innerPredicate.map(this.apply).getOrElse("")
+    s"($v IN $e WHERE $p)"
+  }
+
+  private def backtick(txt: String) = {
+    val c = txt.toCharArray
+    val needsBackticks = !(Character.isJavaIdentifierStart(c(0)) && c.tail.forall(Character.isJavaIdentifierPart))
```

Comment:
What if `txt` is empty?

Reply:
This is used for identifiers. Don't think they can be empty.